src/lib/include/x86_swizzle.h

```diff
@@ -38,25 +47,25 @@ inline __m256d ShuntFirst2Last(const __m256d input) {
 // width of 4, checking that we dont read off the end of ScalarT.
 template <typename VectorT>
 inline VectorT SafeIdxLoad4(const VectorToScalarT<VectorT> *source,
-                            const int idx,
-                            const VectorToScalarT<VectorT> *end) {
+                            std::size_t idx) {
   static_assert(ValuesPerPack<VectorT> == 4,
                 "can only use to load into SIMD datatype of width 4");
   VectorT tmp;
-  if (distopia_likely(source + idx + ValuesPerPack<VectorT> < end)) {
-    // load as is, no overflow, likely path
-    tmp = loadu_p<VectorT>(&source[idx]);
-  } else {
-    // load offset by one, unlikely path
+  if (distopia_likely(idx != 0)) {
+    // not the first, so load starting from value before
     tmp = loadu_p<VectorT>(&source[idx - 1]);
-    tmp = ShuntFirst2Last(tmp);
+  } else {
+    // the first value, so can't go to previous
+    tmp = loadu_p<VectorT>(&source[idx]);
+    tmp = ShuntLast2First(tmp);
   }
   return tmp;
 }
 
 // transforms xyz coordinates from AOS to SOA
-inline void Deinterleave4x3(const __m128 a, const __m128 b, const __m128 c,
-                            const __m128 d, __m128 &x, __m128 &y, __m128 &z) {
+inline void Deinterleave4x3(__m128 a, __m128 b, __m128 c, __m128 d, __m128 &x,
```

Comment:
Is there a reason to remove const?

Reply:
I think it's because the macro below didn't have it.  I think I had to rewrite the 4x3 transpose because we're now keeping the last 3 values on a vector rather than first 3, but then I'm not sure why I kept the rest of the function.  I probably started changing this then got distracted...